frontend/modules/tree/treeProcessing/processing.ts

```diff
@@ -0,0 +1,431 @@
+import { memoize, intersection, keys, pickBy, uniqBy } from 'lodash';
+import { SELECTION_STATES, VISIBILITY_STATES, NODE_TYPES } from '../../../constants/tree';
+
+export class Processing {
+	public nodesList = [];
+	public nodesIndexesMap = {};
+	public defaultVisibilityMap = {};
+	public meshesByModelId = {};
+	public nodesBySharedIdsMap = {};
+	public selectionMap = {};
+	public visibilityMap = {};
+
+	constructor(data) {
+		const {
+			nodesList, nodesIndexesMap, defaultVisibilityMap,
+			meshesByModelId, nodesBySharedIdsMap, visibilityMap, selectionMap
+		} = data;
+		this.nodesList = nodesList;
+		this.nodesIndexesMap = nodesIndexesMap;
+		this.defaultVisibilityMap = defaultVisibilityMap;
+		this.meshesByModelId = meshesByModelId;
+		this.visibilityMap = visibilityMap;
+		this.selectionMap = selectionMap;
+		this.nodesBySharedIdsMap = nodesBySharedIdsMap;
+	}
+
+	public get selectedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
+			return selectionState !== SELECTION_STATES.UNSELECTED;
+		}));
+	}
+
+	public get highlightedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
+			return selectionState === SELECTION_STATES.SELECTED;
+		}));
+	}
+
+	public get invisibleNodesIds() {
+		return keys(pickBy(this.visibilityMap, (selectionState) => {
+			return selectionState === VISIBILITY_STATES.INVISIBLE;
+		}));
+	}
+
+	public clearCurrentlySelected = () => {
+		const selectedNodesIds = this.selectedNodesIds;
+		for (let index = 0, size = selectedNodesIds.length; index < size; index++) {
+			this.selectionMap[selectedNodesIds[index]] = SELECTION_STATES.UNSELECTED;
```

Comment:
✅

Reply:
Selectors are caching the data using shallow compare of theirs args e.g

```
export const selectFullySelectedNodesIds = createSelector(
	selectTreeProccessing, selectDataRevision, (treeProcessingData) => treeProcessingData.fullySelectedNodesIds || []
);
```
This selector will call `fullySelectedNodesIds` getter only if the reference of `selectTreeProccessing` or the value of `selectDataRevision` will have changed. Reference of `selectTreeProccessing` never changes, so the selector will return "fresh" data only if it will get new data revision.

I think the only thing that we can do yet is splitting `selectionMap` and `visibilityMap` into smaller, separate maps e.g `selectedNodes`, `parentOfUnselectedNodes`, 'visibleNodes`, `parentOfInvisibleNodes`, The downside of this solution is that we will need to rewrite each place which contains these maps.

Let me know if I've explained something unclearly and what you think about that. ;)